[WIP] windows initial support

ZeeWanderer opened the pull request:
SO far i had gotten the tests to run but they fail with

```
%%% edb_dap_attach_SUITE: 
%%% edb_dap_attach_SUITE ==> test_attaches_if_node_is_alive: FAILED
%%% edb_dap_attach_SUITE ==> {timeout,{gen_server,call,
                     [<0.1584.0>,
                      #{command => <<"initialize">>,type => request,
                        arguments => #{adapterID => <<"edb for BSCode">>}}]}}
```

```
[2025-07-23T19:54:24.166000+03:00] [info] Application: compiler. Started at: nonode@nohost. [application_controller:info_started/2 L2117] <0.45.0>
[2025-07-23T19:54:24.178000+03:00] [info] Application: crypto. Started at: nonode@nohost. [application_controller:info_started/2 L2117] <0.45.0>
[2025-07-23T19:54:24.179000+03:00] [info] Application: syntax_tools. Started at: nonode@nohost. [application_controller:info_started/2 L2117] <0.45.0>
[2025-07-23T19:54:24.186000+03:00] [info] Supervisor: {local,edb_core_sup}. Started: id=edb_node_monitor,pid=<0.91.0>. [supervisor:report_progress/3 L2240] <0.90.0>
[2025-07-23T19:54:24.186000+03:00] [info] Application: edb_core. Started at: nonode@nohost. [application_controller:info_started/2 L2117] <0.45.0>
[2025-07-23T19:54:24.190000+03:00] [info] Supervisor: {local,edb_sup}. Started: id=edb_dap_thread_id_mappings,pid=<0.95.0>. [supervisor:report_progress/3 L2240] <0.94.0>
[2025-07-23T19:54:24.190000+03:00] [info] Supervisor: {local,edb_sup}. Started: id=edb_dap_frame_id_mappings,pid=<0.96.0>. [supervisor:report_progress/3 L2240] <0.94.0>
[2025-07-23T19:54:24.190000+03:00] [info] Supervisor: {local,edb_sup}. Started: id=edb_dap_vars_ref_mappings,pid=<0.97.0>. [supervisor:report_progress/3 L2240] <0.94.0>
[2025-07-23T19:54:24.192000+03:00] [info] Supervisor: {local,edb_sup}. Started: id=edb_dap_server,pid=<0.98.0>. [supervisor:report_progress/3 L2240] <0.94.0>
[2025-07-23T19:54:24.194000+03:00] [info] Supervisor: {local,edb_sup}. Started: id=edb_dap_transport,pid=<0.99.0>. [supervisor:report_progress/3 L2240] <0.94.0>
[2025-07-23T19:54:24.194000+03:00] [info] Application: edb. Started at: nonode@nohost. [application_controller:info_started/2 L2117] <0.45.0>
```

i assume it's not supposed to be `nonode@nohost`, doesn't show up in `epmd -names` either.

Any advice?

ZeeWanderer committed:
windows pt0

jcpetruzza commented:
First a quick sanity-check question: 
- Are the tests in `edb_SUITE` passing for you? 
  - If not, since you are working on the `main` branch, are you using vanilla OTP-28 or have you built your own from the git submodule in the `otp/` directory? (if the former, please use the `otp-28.0` branch, as you may otherwise encounter test failures)

> i assume it's not supposed to be nonode@nohost, doesn't show up in epmd -names either

I think this is fine. What you are seeing is the DAP server starting; it talks with the client (ie the IDE) via stdin/stdout, so it doesn't need distribution (if the ide later tells it to attach to a running node, it will start distribution, using short or long names, depending on the node it needs to attach to).

> {timeout,{gen_server,call,
>                     [<0.1584.0>,
>                       #{command => <<"initialize">>,type => request,
>                         arguments => #{adapterID => <<"edb for BSCode">>}}]}

This is strange. The "initialize" request is the first thing the client sends so if there's no reply, it means like the very basic communication is failing. According to the [spec](https://microsoft.github.io/debug-adapter-protocol/overview)

> The base protocol exchanges messages that consist of a header and a content part (comparable to HTTP). The header and content part are separated by a \r\n (carriage return, line feed).

So wondering if could it be related to that? E.g. Erlang replacing \r\n in strings by \n or something?

To debug this further you could perhaps enable debug logging?

ZeeWanderer commented:
The problem turned out twofold. Should use `escript` directly instead of `edb.cmd` because that introduces some sort of stdin/stdout layering issue, and you were right. On Windows `\r` is stripped out because the mode is inherited and apparently specifying `binary` does nothing if the inherited mode is `_O_TEXT`. The simplest solution is to relax `<<"\r\n\r\n">>` to `[<<"\r\n\r\n">>, <<"\n\n">>]`. When relaxed, only 56 tests fail.

ZeeWanderer committed:
windows pt1

ZeeWanderer committed:
[edb_dap_configuration_phase_SUITE] fix

ZeeWanderer commented:
I am having a problem with code:get_debug_info/1, it does not return calls. I checked the otp submodule `otp-28.0` branch you recommended i use, and it has no difference from vanilla OPT28.0, which i am using. Did i misunderstand something?

ZeeWanderer committed:
[edb_server_code] make get_debug_info more resilient

ZeeWanderer commented:
Now only 3 tests relying on calls in debug info fail, and the ones needing `wa_test_init` all get skipped

ZeeWanderer committed:
[edb] fix epmd resolution on windows

ZeeWanderer committed:
[edb_gatekeeper] ???

I tested edb with ELP and it turned out that on windows there are string escaping shenanigans. Had to twist the code quite a bit to lessen the amount of things that forcefully get escaped in a manner unsupported by windows powershell.

ZeeWanderer commented:
Can anyone please review [c0173c3](https://github.com/WhatsApp/edb/pull/7/commits/c0173c3e937c5ff6e7511437ae5b9f1dfdd923dd)? I had to twist it quite a bit to get it to work.

ZeeWanderer committed:
[edb_server] don suspend dead pid

ZeeWanderer committed:
[edb_bootstrap] add_project_paths

for some reason on windows with rebatr3 edb wails with modules not found if project ebin is not added.

ZeeWanderer commented:
Still cant hit a breackpoint on windows with vscode + elp + edb but at the very least it interfaces correctly and loads the project modules.

robertoaloi commented:
@ZeeWanderer I'd say let's first get the Windows-specific bits which are not affecting the regular mechanism (e.g. the carriage return / new line changes in one PR, the escript-related ones in another one). Then we can look at the core changes (`edb_server`, `edb_server_code`). The rebar3 changes can be also separate and they probably need some generalization (e.g. one could use different *profiles*?)
